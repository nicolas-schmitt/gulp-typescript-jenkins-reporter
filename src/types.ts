import { TypeScriptError } from 'gulp-typescript/release/reporter';

export interface FormattedError {
    xml: string;
    path: string;
}

export type TypescriptErrorPosition = NonNullable<TypeScriptError['startPosition']>;

export interface ReportOptions {
    sort: boolean;
    filename: string;
    pathBase: string;
    pathPrefix: string;
}
