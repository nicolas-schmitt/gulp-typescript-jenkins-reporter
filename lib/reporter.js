"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const lodash_1 = require("lodash");
const path_1 = __importDefault(require("path"));
const upath_1 = require("upath");
const formatter_1 = require("./formatter");
function getDefaultOptions(partialOptions) {
    return lodash_1.defaults({}, partialOptions, {
        sort: false,
        filename: 'pmd.xml',
        pathBase: '',
        pathPrefix: '',
    });
}
function report(partialOptions = {}) {
    let errorBuffer = [];
    const options = getDefaultOptions(partialOptions);
    return {
        error: (error) => {
            const file = error === null || error === void 0 ? void 0 : error.file;
            if (file) {
                errorBuffer.push({
                    xml: formatter_1.formatError(error),
                    path: getReportedFilePath(options, file),
                });
            }
        },
        finish: () => {
            if (options.sort) {
                errorBuffer = lodash_1.sortBy(errorBuffer, 'path');
            }
            const content = formatter_1.formatStream(errorBuffer);
            fs_1.default.promises.writeFile(options.filename, content).catch(error => {
                // tslint:disable-next-line:no-console
                console.error(error);
            });
        },
    };
}
exports.report = report;
function getReportedFilePath(partialOptions, vinyl) {
    const options = getDefaultOptions(partialOptions);
    let result = upath_1.normalize(vinyl.path);
    if (options) {
        if (options.pathBase) {
            const index = result.indexOf(options.pathBase);
            if (index > 0) {
                result = result.substr(index + options.pathBase.length);
            }
        }
        if (options.pathPrefix) {
            result = path_1.default.join(options.pathPrefix, result);
        }
    }
    if (path_1.default.sep === path_1.default.win32.sep) {
        result = result.replace(/\//, path_1.default.win32.sep);
    }
    return result;
}
exports.getReportedFilePath = getReportedFilePath;
