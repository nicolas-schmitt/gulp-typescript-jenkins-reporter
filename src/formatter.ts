import { TypeScriptError } from 'gulp-typescript/release/reporter';
import { escape, groupBy } from 'lodash';
import { DiagnosticCategory, flattenDiagnosticMessageText } from 'typescript';
import { FormattedError, TypescriptErrorPosition } from './types';

const DefaultPosition: TypescriptErrorPosition = {
    position: 0,
    line: 0,
    character: 0,
};

export function formatStream(errors: Array<FormattedError>): string {
    const xml = Object.entries(groupBy(errors, 'path'))
        .map(([path, fileErrors]) => {
            return formatFile(path, fileErrors);
        })
        .join('');

    return `<?xml version="1.0" encoding="utf-8"?>\n<pmd version="4.2.5">${xml}</pmd>`;
}

export function formatFile(fileName: string, errors: Array<FormattedError>) {
    if (!errors.length) {
        return '';
    }

    return `<file name="${fileName}">${errors.map(error => error.xml).join('')}</file>`;
}

export function formatError(error: TypeScriptError): string {
    const start = error.startPosition || DefaultPosition;
    const end = error.endPosition || DefaultPosition;
    const message = escape(flattenDiagnosticMessageText(error.diagnostic.messageText, '\n'));

    return formatTagWithAttributes(
        'violation',
        {
            beginline: start.line,
            begincolumn: start.character,
            endline: end.line,
            endcolumn: end.character,
            priority: error.diagnostic.category,
            rule: 'TS' + error.diagnostic.code,
        },
        message
    );
}

function formatTagWithAttributes(
    tagName: string,
    attributes: { [key: string]: string | number | DiagnosticCategory },
    children?: string
): string {
    const formattedAttributeList = Object.entries(attributes)
        .filter(([_, value]) => value !== null && value !== undefined)
        .map(([name, value]) => `${name}="${value}"`)
        .join(' ');

    const formattedAttributes = formattedAttributeList ? ' ' + formattedAttributeList : '';

    return `<${tagName}${formattedAttributes}>${children}</${tagName}>`;
}
