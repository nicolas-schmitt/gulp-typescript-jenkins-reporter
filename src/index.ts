import { report } from './reporter';

export default report;
