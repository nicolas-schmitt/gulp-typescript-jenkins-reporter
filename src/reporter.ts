import { promises } from 'fs';
import { Reporter, TypeScriptError } from 'gulp-typescript/release/reporter';
import { VinylFile } from 'gulp-typescript/release/types';
import { defaults, sortBy } from 'lodash';
import path from 'path';
import { normalize as normalizePath } from 'upath';
import Formatter from './formatter';
import { FormattedError, ReportOptions } from './types';
const { writeFile } = promises;


function getDefaultOptions(partialOptions: Partial<ReportOptions> | undefined): ReportOptions {
    return defaults({}, partialOptions, {
        sort: false,
        filename: 'pmd.xml',
        pathBase: '',
        pathPrefix: '',
    });
}

export function report(partialOptions: Partial<ReportOptions> = {}): Reporter {
    let errorBuffer: Array<FormattedError> = [];

    const options = getDefaultOptions(partialOptions);

    const formatter = new Formatter();

    return {
        error: (error: TypeScriptError) => {
            const file = error?.file;
            if (file) {
                errorBuffer.push({
                    xml: formatter.formatError(error),
                    path: getReportedFilePath(options, file),
                });
            }
        },
        finish: () => {
            if (options.sort) {
                errorBuffer = sortBy(errorBuffer, 'path');
            }

            const content = formatter.formatStream(errorBuffer);
            writeFile(options.filename, content).catch(error => {
                console.error(error);
            });
        },
    };
}

export function getReportedFilePath(partialOptions: Partial<ReportOptions>, vinyl: VinylFile): string {
    const options = getDefaultOptions(partialOptions);
    let result = normalizePath(vinyl.path);

    if (options) {
        if (options.pathBase) {
            const index = result.indexOf(options.pathBase);
            if (index > 0) {
                result = result.substr(index + options.pathBase.length);
            }
        }

        if (options.pathPrefix) {
            result = path.join(options.pathPrefix, result);
        }
    }

    if (path.sep === path.win32.sep) {
        result = result.replace(/\//, path.win32.sep);
    }

    return result;
}
