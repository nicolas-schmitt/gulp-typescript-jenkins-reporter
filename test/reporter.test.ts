import fs from 'fs';
import gulp from 'gulp';
import ts from 'gulp-typescript';
import { getReportedFilePath, report } from '../src/reporter';
import { parseAndValidate } from './util';

jest.setTimeout(15000);

describe('reporter', () => {
    const tsFileShim: any = {
        path: '/home/ubuntu/workspace/src/file.js',
    };

    function createReport(path?: string) {
        const project = ts.createProject({});
        return new Promise(resolve => {
            gulp.src('./test/mocks/dirty.ts')
                .pipe(project(report({ filename: path })))
                .on('finish', resolve)
                // tslint:disable-next-line:no-empty
                .on('error', () => {});
        });
    }

    it('report should create a valid xml report', async () => {
        const reportPath = './pmd.xml';
        await createReport(reportPath);
        const contents = await fs.promises.readFile(reportPath);

        expect(parseAndValidate(contents.toString())?.documentElement).toMatchSnapshot();

        await fs.promises.unlink(reportPath);
    });

    describe('getReportedFilePath', () => {
        it("shouldn't rebase the path when options.pathBase is undefined", () => {
            const actual = getReportedFilePath({}, tsFileShim);
            expect(actual).toBe(tsFileShim.path);
        });

        it("shouldn't rebase the path when options.pathBase is empty", () => {
            const actual = getReportedFilePath({ pathBase: '' }, tsFileShim);
            expect(actual).toBe(tsFileShim.path);
        });

        it('should rebase the path when options.pathBase is set', () => {
            const actual = getReportedFilePath({ pathBase: '/workspace' }, tsFileShim);
            expect(actual).toBe('/src/file.js');
        });

        it('should prepend the path when options.pathPrefix is set', () => {
            const actual = getReportedFilePath({ pathPrefix: '/project' }, tsFileShim);
            expect(actual).toBe('/project' + tsFileShim.path);
        });

        it('should prepend and rebase the path when both options are set', () => {
            const actual = getReportedFilePath(
                {
                    pathBase: '/workspace',
                    pathPrefix: '/project',
                },
                tsFileShim
            );
            expect(actual).toBe('/project/src/file.js');
        });
    });
});
