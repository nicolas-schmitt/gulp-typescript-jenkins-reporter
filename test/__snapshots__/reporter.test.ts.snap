// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reporter report should create a valid xml report 1`] = `
<pmd
  version="4.2.5"
>
  <file
    name="/test/mocks/dirty.ts"
  >
    <violation
      begincolumn="14"
      beginline="12"
      endcolumn="18"
      endline="12"
      priority="1"
      rule="TS2345"
    >
      Argument of type 'number' is not assignable to parameter of type 'string'.
    </violation>
    <violation
      begincolumn="13"
      beginline="13"
      endcolumn="20"
      endline="13"
      priority="1"
      rule="TS2551"
    >
      Property 'weaponn' does not exist on type 'Peon'. Did you mean 'weapon'?
    </violation>
    <violation
      begincolumn="13"
      beginline="20"
      endcolumn="19"
      endline="20"
      priority="1"
      rule="TS2341"
    >
      Property 'weapon' is private and only accessible within class 'Orc'.
    </violation>
    <violation
      begincolumn="10"
      beginline="28"
      endcolumn="11"
      endline="28"
      priority="1"
      rule="TS2430"
    >
      Interface 'B' incorrectly extends interface 'A'.
  Types of property 'prop' are incompatible.
    Type 'string' is not assignable to type 'number'.
    </violation>
  </file>
</pmd>
`;
